clustercontroller-reindexer/src/main/java/ai/vespa/reindexing/Reindexer.java

```diff
@@ -0,0 +1,235 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package ai.vespa.reindexing;
+
+import ai.vespa.reindexing.Reindexing.Status;
+import ai.vespa.reindexing.ReindexingCurator.ReindexingLockException;
+import com.yahoo.document.DocumentType;
+import com.yahoo.document.select.parser.ParseException;
+import com.yahoo.documentapi.DocumentAccess;
+import com.yahoo.documentapi.ProgressToken;
+import com.yahoo.documentapi.VisitorControlHandler;
+import com.yahoo.documentapi.VisitorParameters;
+import com.yahoo.documentapi.VisitorSession;
+import com.yahoo.documentapi.messagebus.protocol.DocumentProtocol;
+import com.yahoo.vespa.curator.Lock;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Phaser;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Logger;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.logging.Level.FINE;
+import static java.util.logging.Level.INFO;
+import static java.util.logging.Level.WARNING;
+import static java.util.stream.Collectors.joining;
+
+/**
+ * Progresses reindexing efforts by creating visitor sessions against its own content cluster,
+ * which send documents straight to storage — via indexing if the documenet type has "index" mode.
+ * The {@link #reindex} method blocks until shutdown is called, or until no more reindexing is left to do.
+ *
+ * @author jonmv
+ */
+public class Reindexer {
+
+    private static final Logger log = Logger.getLogger(Reindexer.class.getName());
+
+    private final Cluster cluster;
+    private final Map<DocumentType, Instant> ready;
+    private final ReindexingCurator database;
+    private final DocumentAccess access;
+    private final Clock clock;
+    private final Phaser phaser = new Phaser(2); // Reindexer and visitor.
+
+    private Reindexing reindexing;
+    private Status status;
+
+    public Reindexer(Cluster cluster, Map<DocumentType, Instant> ready, ReindexingCurator database,
+                     DocumentAccess access, Clock clock) {
+        for (DocumentType type : ready.keySet())
+            cluster.bucketSpaceOf(type); // Verifies this is known.
+
+        this.cluster = cluster;
+        this.ready = new TreeMap<>(ready); // Iterate through document types in consistent order.
+        this.database = database;
+        this.access = access;
+        this.clock = clock;
+    }
+
+    /** Tells this to stop reindexing at its leisure. */
+    public void shutdown() {
+        phaser.forceTermination();
```

Comment:
For those of us not too familiar with these newfangled and fancy `Phaser` doohickies, a comment on how this ends up transitively aborting the active visitor session would be nice 🦖

Reply:
Of course!